### PR TITLE
Clarify producer span relationships and add a messaging example for a batch publishing scenario

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes open-telemetry/opentelemetry-specification#1466

## Changes

Based on recent changes from #284, this adds a messaging example illustrating a batch-publishing scenario involving "Publish" and "Create" spans.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

**Review by joaopgrassi — DISMISSED** [review]

**Review by AlexanderWert — APPROVED** [review]

File: docs/messaging/messaging-spans.md [hunk 1]
```diff
@@ -512,6 +513,51 @@ flowchart LR;
...
+  linkStyle 0,1,2,3 color:green,stroke:green
+```
+
+| Field or Attribute | Span Publish | Span Create A | Span Create B | Span Receive A | Span Receive B |
+|-|-|-|-|-|-|
+| Span name | `Q publish` | `Q create` | `Q create` | `Q receive` | `Q receive` |
+| Parent |  | Span Publish | Span Publish | | |
+| Links |  |  |  | Span Create A | Span Create B |
+| SpanKind |  | `PRODUCER` | `PRODUCER` | `CONSUMER` | `CONSUMER` |
+| Status | `Ok` | `Ok` | `Ok` | `Ok` | `Ok` |
```
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:553]: nit: we never set status to `OK` in other semconv, but leave it unset, so maybe we should just remove this row since it's not strongly related to the example?

File: docs/messaging/messaging-spans.md [hunk 2]
```diff
@@ -512,6 +513,51 @@ flowchart LR;
...
+  end
+  subgraph CONSUMER1
+  direction TB
+  D1[Span Receive A]
+  end
+  subgraph CONSUMER2
+  direction TB
+  D2[Span Receive B]
+  end
+  P-- parent -->CA;
```
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:537]: I'd really prefer to change it to a more realistic and advanced example with links. 
The problem with parenting is that instrumentation assumes that messages are created and published in the same ambient context.

In reality API looks like [`send(ProducerRecord<K,V> record,
                                                        Callback callback)`](https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html) and instrumentation has no idea in which context messages were created.

On Azure, we even provide buffered producer for EventHubs which publishes messages in a background thread.

As a result, in a general case, the only reasonable instrumentation point for `create` operation is message creation (in the case of Azure, we do it when a message is added to the batch).

It means that publish operation would start after creation and cannot be a parent for creation.

Long story short, I'd prefer to have links here instead of parent and also in the table below.

Comment (lmolkova) [reply docs/messaging/messaging-spans.md:537]: it seems kafka is even more nuanced - 
https://stackoverflow.com/questions/58492689/how-to-send-batched-data-with-spring-kafka-producer.

`send` is not a send, but rather 'add-to-batch' and callback when actually sent. So this example stands as is for Kafka. But I'm a bit worried it creates a wrong impression for the general case of batch sending. Let me do a small research and let's quickly discuss on the messaging SIG on Thursday.

Comment (lmolkova) [reply docs/messaging/messaging-spans.md:535]: Did a bit of a research on how batch publish APIs look like:

[AmazonSQS](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sqs/AmazonSQS.html#sendMessageBatch-com.amazonaws.services.sqs.model.SendMessageBatchRequest-), [RocketMQ](https://rocketmq.apache.org/docs/4.x/producer/05message4/), Azure [EventHubs](https://javadoc.io/doc/com.azure/azure-messaging-eventhubs/latest/com/azure/messaging/eventhubs/EventHubBufferedProducerClient.html), and [EventGrid](https://javadoc.io/doc/com.azure/azure-messaging-eventgrid/latest/com/azure/messaging/eventgrid/EventGridPublisherClient.html) batch producer API take a list of messages. With such APIs it's hard to assume that publish is called in the same ambient context as message creation.

Also, thinking more about buffered client like Kafka (also Pulsar and GCP PubSub), even there parent relationships would be wrong. They buffer messages until buffer is full and then start publishing them. In this case by the time publishing starts all the ambient context in which messages were created is lost. 
image

Comment (lmolkova) [reply docs/messaging/messaging-spans.md:535]: Given this,
- Kafka or other buffering client that publishes [asynchronously](https://docs.confluent.io/kafka-clients/java/current/overview.html#asynchronous-writes) should always create messages spans in the context `send` is called, before actual publishing starts.
- Clients that publish one message [sychronously](https://docs.confluent.io/kafka-clients/java/current/overview.html#synchronous-writes) may use parent-child relationships. 
- But sync client that publish a batch should not assume messages are created in the same context in which publish is called. Therefore it's always better to create a message span when a message is created, not when it's published.

I think it would be easier to implement and visualize if we recommended links in all cases and changed

https://github.com/open-telemetry/semantic-conventions/blob/a4c60f7464eb84d581819b9d0181d9c0433f471b/docs/messaging/messaging-spans.md?plain=1#L243-L244

to something like `"Create" span SHOULD be created when message is created or passed to the client library (or other component) responsible for publishing. "Publish" span SHOULD link to the message creation contexts.`

Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: I adapted the example (and the wording for producer span relationships) according to our discussions in the workgroup.

@lmolkova @alevenberg @joaopgrassi Please have another look.

File: docs/messaging/messaging-spans.md [hunk 3]
```diff
@@ -512,6 +512,50 @@ flowchart LR;
...
+  end
+  CA-. link .-P;
+  CB-. link .-P;
+  CA-. link .-D1;
+  CB-. link .-D2;
+
+  classDef normal fill:green
+  class P,CA,CB,D1,D2 normal
+  linkStyle 0,1,2,3 color:green,stroke:green
+```
```
Comment (alevenberg) [inline docs/messaging/messaging-spans.md:544]: I added some colors to the diagram so it's easier to distinguish the links

```suggestion
```mermaid
flowchart TD;
  subgraph CONSUMER2
  direction LR
  RB[Receive B]
  end
  subgraph CONSUMER1
  direction LR
  RA[Receive A]
  end
  subgraph PRODUCER
  direction LR
  CA[Create A]
  CB[Create B]
  P[Publish]
  end
  %% Link 0
  CA-. link .-P;  
  %% Link 1
  CB-. link .-P;
  %% Link 2
  CA-. link .-RA;
  %% Link 3
  CB-. link .-RB;

  %% Style the nodes and corresponding links
  %% producer links and nodes
  classDef producer fill:green
  class P,CA,CB producer
  linkStyle 0,1 color:green,stroke:green

  %% consumer1 links and nodes with light blue
  classDef consumer1 fill:#9eecff,color:black
  class RA consumer1
  linkStyle 2 color:#9eecff,stroke:#9eecff

  %% consumer2 links and nodes with dark blue
  classDef consumer2 fill:#032a61
  class RB consumer2
  linkStyle 3 color:#032a61,stroke:#032a61
```
```

File: docs/messaging/messaging-spans.md (hunk 2 above)
Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: Trying to prototype this, I came across some problems and questions.

Currently, we require a "publish" span to always be present on producer side. "create" spans can be present, and are especially important for batch publishing scenarios.

Furthermore, as a result of recent discussions, "publish" spans should cover the operation of the SDK sending messages to the intermediary, and should also be able to capture the duration and success of this operation.

This leads to some open questions:
1. When instrumenting an abstraction layer like JMS or MassTransit, this is hard to accomplish. The "publish" span would need to be created by messaging SDKs used under the cover, however, that's often not possible and thus makes it hard for these abstraction layers to provide consistent "publish" spans. As an alternative, "publish" spans could be created to cover `publish` calls of the API.
2. In case a customer sends a single message without creation context attached, the context of the "publish" operatino would be attached to this message. This context doesn't correspond to the API `publish` call, but possibly to some background thread taking and sending messages from a queue. With this setup, the creation context might be of limited use, as it doesn't correspond to the user context in which the message was created and passed via the API `publish` call, but to the SDK context for sending messages in the background. As an alternative, "create" spans can be created.

One problem I see with the "publish" span modelling the SDK-internal operation of publishing messages is, that it is of limited use when trying to track the origin and flow of a message. It is useful for seeing what's going inside an SDK, but I think in most cases users don't care much whether SDKs batch messages under the cover, but they rather want to correlate producer and consumer traces (and not traces relating to SDK internals).

File: docs/messaging/messaging-spans.md [hunk 4]
```diff
@@ -512,6 +512,50 @@ flowchart LR;
...
+
+| Field or Attribute | Span Publish | Span Create A | Span Create B | Span Receive A | Span Receive B |
```
[the remark this answers is not in the document]
Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: > * Changed the order so it's Create A / Create B / Publish / Receive A / Receive B, since that's the order they would be created in

Incorporated.

> * Made the publish span a Client SpanKind, added a Producer span kind for each create span

Done. In this scenario, "publish" can either be an internal or a client span.

> * Removed the parent spans and added bidirectional links between the publish span to each of the creates

Removed the parent spans. However, I didn't add bi-directional links. We should denote here which span is supposed to add the link.

> * Added bidrectional links from the create to receive spans

I didn't apply that, see the reason above.

> * Added the messaging.batch.message_count attribute

Done.

File: docs/messaging/messaging-spans.md [hunk 5]
```diff
@@ -216,8 +217,8 @@ SHOULD be set according to the following table, based on the operation a span de
 
 | Operation name | Span kind|
 |----------------|-------------|
-| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
 | `create`       | `PRODUCER` |
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
```
[the remark this answers is not in the document]
Comment (pyohannes) [reply docs/messaging/messaging-spans.md:221]: If it's not a `PRODUCER` span, we shouldn't make any assumptions about the span kind. It could be a `CLIENT` span, however, there could also be transport-level `CLIENT` spans, in which case the "publish" span would be `INTERNAL`.

[the remark this answers is not in the document]
Comment (alevenberg) [reply docs/messaging/messaging-spans.md:221]: Ok sounds good.

File: docs/messaging/messaging-spans.md (hunk 2 above)
Comment (joaopgrassi) [reply docs/messaging/messaging-spans.md]: > When instrumenting an abstraction layer like JMS or MassTransit, this is hard to accomplish. The "publish" span would need to be created by messaging SDKs used under the cover, however, that's often not possible and thus makes it hard for these abstraction layers to provide consistent "publish" spans. As an alternative, "publish" spans could be created to cover publish calls of the API.

If I understand it right, you mean the abstraction (e.g. MassTransit) exposes a `publish` methods developers use. But under the hood, it invokes another publish from the actual messaging library that actually does the work? And in some cases, the abstraction does not "send right away" but can do all sort of other things (internally caching messages and send in a batch and etc), so the "publish" of the abstraction is not really useful/measure the real API call?

Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: Another problem I came across with the "publish" span modelling the SDK-internal operation of publishing is related to add instrumentation on top of SDKs that aren't instrumented.

For example, that's how a message is published when using the .NET Apache Kafka client library:
```csharp
public static void handler(DeliveryReport<Null, string>)
{
    ...
}

var message = new Message<Null, string> {
  Value = "hello world"
};

producer.Produce("my-topic", message, handler);
```

This is quite a common API pattern offered by messaging SDKs. If such an SDK doesn't offer built-in instrumentation, creating a "publish" that covers the duration until the message is published as well as the publishing status is quite involved. It would require starting a span before the call to `Produce`, decouple the span from the stack of active contexts without ending it, caching it, retrieving the right cached span during a call to the `handler`, and there end it with the status extracted from the `DeliveryReport`.

Comment (joaopgrassi) [reply docs/messaging/messaging-spans.md]: Not great for sure. But I guess we, as conventions shouldn't worry much about or? From the point that instrumentations would then add instrumentation and things would then work, or you think this assumption is also wrong/not good?

Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: I'll put it on the agenda for tomorrow's workgroup meeting.

Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: After some research and discussions, I'm convinced that @lmolkova's suggestions are feasible and I reworded the paragraph about "Create" and "Publish" spans accordingly:

> "Create" spans MAY be created when a message is created or passed to the client
library or other component responsible for publishing.  A single "Create" span
SHOULD account only for a single message.  "Publish" spans SHOULD be created
for operations of sending or publishing a message to an intermediary. A single
"Publish" span can account for a single message, or for multiple messages (in
the case of sending messages in batches).  "Publish" spans SHOULD link to the
message creation contexts.

@lmolkova @alevenberg, Please have another look.

File: docs/messaging/messaging-spans.md (hunk 3 above)
Comment (pyohannes) [reply docs/messaging/messaging-spans.md:554]: I'd prefer using too many colors, as it might confuse readers.

To make the meaning of colors clear, I added [this paragraph from OTEP 220](https://github.com/open-telemetry/oteps/blob/main/text/trace/0220-messaging-semantic-conventions-span-structure.md#examples) to the beginning of the "Examples" section:

> This section contains a list of examples illustrating the use of the conventions outlined above. Green boxes denote spans that are required to exist in order to conform to those conventions. Other boxes denote spans that are not required and covered by the conventions, but are hopefully helpful in understanding how messaging spans can be integrated into an overall trace flow. Solid arrows denote parent/child relationships, dotted arrows denote link relationships.

**Review by hongalex — APPROVED** [review]

File: docs/messaging/messaging-spans.md (hunk 2 above)
Comment (joaopgrassi) [reply docs/messaging/messaging-spans.md]: @pyohannes  So giving this problem above https://github.com/open-telemetry/semantic-conventions/pull/509#discussion_r1447531471, with the new wording this still occurs IIUC. I remember that we discussed in the meeting that the process of instrumenting that is involved, but not impossible, thus we are OK with it, giving it still follows our conventions?

File: docs/messaging/messaging-spans.md [hunk 6]
```diff
@@ -236,20 +237,23 @@ interpret linked traces without the need for additional semantic hints.
...
 
-"Publish" spans SHOULD be created for operations of providing messages for
-sending or publishing to an intermediary. A single "Publish" span can account
-for a single message, or for multiple messages (in the case of providing
-messages in batches). "Create" spans MAY be created. A single "Create" span
-SHOULD account only for a single message. "Create" spans SHOULD either be
-children or links of the related "Publish" span.
+"Create" spans MAY be created when a message is created or passed to the client
+library or other component responsible for publishing.  A single "Create" span
+SHOULD account only for a single message.  "Publish" spans SHOULD be created
```
Comment (joaopgrassi) [inline docs/messaging/messaging-spans.md:242]:
```suggestion
SHOULD account only for a single message.  

"Publish" spans SHOULD be created
```

What if we break the line, so each "block" refers to one "thing"? I think that will improve readability.

**Review by joaopgrassi — APPROVED** [review]

File: docs/messaging/messaging-spans.md (hunk 6 above)
Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: If have no strong opinion, we can do that if people think it increases readibility.

The logic I was following with the paragraphs:
* Paragraph 1: What spans to create
* Paragraph 2: What context to inject
* Paragraph 3: What links to create

File: docs/messaging/messaging-spans.md (hunk 2 above)
Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: @joaopgrassi Yes, we said that instrumenting in this way can be cumbersome, but is certainly possible. It was a consensus that this is acceptable.

**Review by lmolkova — APPROVED** [review]
Please also update the changelog

**Comment by joaopgrassi** [comment]
@pyohannes I think we should be good to merge this right? Can you please move the changelog changes to the new file-based one?

**Comment by pyohannes** [comment]
> @pyohannes I think we should be good to merge this right? Can you please move the changelog changes to the new file-based one?

Done. This should be good to merge.

**Comment by joaopgrassi** [comment]
Can we close the discussions, or you are still waiting for resolution on those? Merge is blocked by them now.
